### PR TITLE
Fix/nil read line

### PR DESCRIPTION
**Pull request by orlando-labs**
Hi, this PR addresses some rare [issues](https://github.com/socketry/protocol-http1/issues/12) when client dies suddenly causing IOError in IO.gets/Async::IO::Stream.read_until methods. [This PR](https://github.com/socketry/async-http/pull/70) for [async-http](https://github.com/socketry/async-http) covers the same error

### CODE REVIEW

**Comment by ioquatix** [comment]
Thanks for this.

Does it make sense to silence this error? Because it's a protocol failure. There is no case where we can be reading a chunked response and the size line is nil or the client stops at that point.

**Comment by ioquatix** [comment]
As an example of where this might fail, imagine you are uploading a large file, on the server side `#read` returns nil so you think you are done but actually it was a network failure. I think this error condition should be transparent to the consumer of the chunked body.

**Comment by orlando-labs** [comment]
Yeah, I got your point as it is correct. But all this time I was resolving the following case:
1. upload file with curl
2. respond with 404 successfully
```
> POST /api/some_path HTTP/1.1
> Host: localhost:9292
> User-Agent: curl/7.61.1
> Accept: */*
> Content-Type: application/octet-stream
> Transfer-Encoding: chunked
> Expect: 100-continue
>
< HTTP/1.1 404 Not Found
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< x-frame-options: SAMEORIGIN
< vary: accept-encoding
< content-length: 0
* HTTP error before end of send, stop sending
```
3. Server raises the aforesaid IOError exception.

So, first of all, the server transmits the response to client. Then client drops the connection. Then the error occurs.

**Comment by orlando-labs** [comment]
So maybe for now we should stand for just checking the lines for nil values.

**Comment by ioquatix** [comment]
Do you think you can help me understand the problem by making a repro script?

**Comment by orlando-labs** [comment]
Yes, I can and would. In several days

**Comment by ioquatix** [comment]
With `config.ru` that always returns 404:

```
cat config.ru | curl --upload-file . -v http://localhost:9494
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9494 (#0)
> PUT / HTTP/1.1
> Host: localhost:9494
> User-Agent: curl/7.64.1
> Accept: */*
> Transfer-Encoding: chunked
> Expect: 100-continue
> 
< HTTP/1.1 404 Not Found
< vary: accept-encoding
< content-length: 0
* HTTP error before end of send, stop sending
< 
* Closing connection 0
```

**Comment by ioquatix** [comment]
It seems to me that in this case curl is not compliant with the specifications, but the server is also not clear what to do.

1. Curl says it's going to upload data.
2. As soon as 404 is received, curl terminates the connection.
3. Curl does not gracefully terminate the stream, because it's not empty, and HTTP/1 provides no way to say "I'm terminating this stream but you didn't receive the data." - sending a single "0" line might make it look like the data was empty rather than it wasn't sent.

That being said, there is a race condition here:

```ruby
# config.ru

app = lambda do |env|
	Async::Task.current.sleep(2)
	[404, [], []]
end

run app
```

```
samuel@Fukurou ~/D/s/f/e/posty (master)> cat config.ru | curl --upload-file . -v http://localhost:9494
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9494 (#0)
> PUT / HTTP/1.1
> Host: localhost:9494
> User-Agent: curl/7.64.1
> Accept: */*
> Transfer-Encoding: chunked
> Expect: 100-continue
> 
* Done waiting for 100-continue
* Signaling end of chunked upload via terminating chunk.
< HTTP/1.1 404 Not Found
< vary: accept-encoding
< content-length: 0
< 
* Connection #0 to host localhost left intact
* Closing connection 0
```

This works correctly because curl buffers the entire file in the right format, and then closes the connection after receiving the 404.

I want to know more about the behaviour of curl in this case.

However, we should not ignore empty length line, but if it's not integer, we should fail (but argument error is wrong exception). Maybe we should introduce `ChunkError` or something.

**Comment by ioquatix** [comment]
In the following example, sending 200 is okay, but sending 404 causes hang on server side:

```
app = lambda do |env|
	log_input_stream = lambda do |stream|
		Console.logger.info(stream, stream.read)
	ensure
		stream.close
	end
	
	[404, {'rack.hijack' => log_input_stream}, nil]
end

run app
```

**Comment by orlando-labs** [comment]
I tried with the following setup:
config.ru
```ruby
run lambda { |env| [404, {}, []] }
```

```
[orlylabs@ai4 app]$ falcon serve -n 1 -b "http://localhost:9292"
```
```
[orlylabs@ai4 app]$ curl -v -H 'Transfer-Encoding: chunked' --data-binary @a_file "http://localhost:9292/give_me_404"
```

This results in
```
11.91s    error: Async::Task [oid=0xb18] [ec=0xb2c] [pid=44675] [2021-04-23 10:46:23 +0000]
               |   ArgumentError: wrong number of arguments (given 1, expected 0)
               |   → /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http1-0.14.0/lib/protocol/http1/body/chunked.rb:59 in `to_i'                                                                                   
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http1-0.14.0/lib/protocol/http1/body/chunked.rb:59 in `read'                                                                                   
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http-0.22.0/lib/protocol/http/body/readable.rb:86 in `each'                                                                                    
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http-0.22.0/lib/protocol/http/body/buffered.rb:47 in `for'                                                                                     
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http-0.22.0/lib/protocol/http/body/readable.rb:81 in `finish'                                                                                  
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/protocol-http-0.22.0/lib/protocol/http/body/reader.rb:52 in `finish'                                                                                    
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/async-http-0.56.1/lib/async/http/protocol/http1/server.rb:107 in `each'                                                                                 
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/async-http-0.56.1/lib/async/http/server.rb:53 in `accept'                                                                                               
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/async-io-1.30.2/lib/async/io/server.rb:32 in `block in accept_each'                                                                                     
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/async-io-1.30.2/lib/async/io/socket.rb:73 in `block in accept'                                                                                          
               |     /home/orlylabs/.bundle/ruby/3.0.0/gems/async-1.29.0/lib/async/task.rb:263 in `block in make_fiber' 
```

**Comment by orlando-labs** [comment]
In my case curl did't started the upload of a file as 404 strikes before end of waiting for 100 continue

**Comment by ioquatix** [comment]
I think it's the fault of HTTP/1 protocol. There is no official way to terminate stream in the case you decide not to send it after sending headers. Sending `0\r\n\r\n` is not acceptable because it might indicate success "empty data/body" rather than "I don't want to send data".

In async-http, you can use socket shutdown, i.e. `IO#close_read` and `IO#close_write` do work correctly. But curl seems to assume that if it don't want to upload, its only option is to close connection.

As a counter point, with HTTP/2 this is a non-issue, and I even tested it with the same as the above.

**Comment by ioquatix** [comment]
> In my case curl did't started the upload of a file as 404 strikes before end of waiting for 100 continue

You need to call sleep before sending 404, for a few seconds, then curl will send body, and it all works correctly.

**Comment by ioquatix** [comment]
By the way, just because server return 404 doesn't mean it won't read body. I'm not sure if this is right behaviour by curl. Maybe the RFC for expect `100-continue` clarifies this, but I feel that curl is wrong here.

**Comment by ioquatix** [comment]
We need to find out if what curl is doing is compliant with HTTP RFC.

**Comment by orlando-labs** [comment]
Ok, i'll try to figure it out with additional tests and research. Anyway, falling down with invalid :to_i arguments looks kinda not beautiful even if it is the client fault :)

File: lib/protocol/http1/connection.rb [hunk 1]
```diff
@@ -171,6 +171,8 @@ def write_headers(headers)
 			
 			def read_line?
 				@stream.gets(CRLF, chomp: true)
+			rescue IOError, Errno::ECONNRESET
```
Comment (ioquatix) [inline lib/protocol/http1/connection.rb:174]: Can you rebase on master. Not sure if we need this on the latest version or not. Thoughts?

**Pull request closed without merging**